clients/unshaded/src/main/java/tachyon/client/block/BufferedBlockInStream.java

```diff
@@ -0,0 +1,244 @@
+  /** Value to set mBufferPos to mark the buffer as invalid. */
+  private static final long INVALID_BUFFER_POS = -1L;
+
+  /** The id of the block to which this instream provides access. */
+  protected final long mBlockId;
+  /** The size in bytes of the block. */
+  protected final long mBlockSize;
+  /** The address of the worker to read the data from. */
+  protected final InetSocketAddress mLocation;
+  /** The block store context which provides block worker clients. */
+  protected final BlockStoreContext mContext;
+
+  /** Internal buffer to improve small read performance. */
+  protected ByteBuffer mBuffer;
+  /** Flag indicating if the stream is closed, can only go from false to true. */
+  protected boolean mClosed;
+  /** Current position of the buffer's next byte, relative to the start of the block */
+  protected long mBufferPos;
+  /** Current position of the stream, relative to the start of the block. */
+  protected long mPos;
+
+  /**
+   * Helper for creating a BufferedBlockInStream. This sets the necessary variables and creates
+   * the initial buffer which is empty and invalid.
+   *
+   * @param blockId block id for this stream
+   * @param blockSize size of the block in bytes
+   * @param location worker address to read the block from
+   */
+  // TODO: Get the block lock here when the remote instream locks at a stream level
+  public BufferedBlockInStream(long blockId, long blockSize, InetSocketAddress location) {
+    mBlockId = blockId;
+    mBlockSize = blockSize;
+    mLocation = location;
+    mBuffer = allocateBuffer();
+    mBufferPos = INVALID_BUFFER_POS; // No data in buffer
+    mClosed = false;
+    mContext = BlockStoreContext.INSTANCE;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+  }
+
+  @Override
+  public int read() throws IOException {
+    checkIfClosed();
+    if (remaining() == 0) {
+      close();
+      return -1;
+    }
+    if (remainingInBuffer() == 0) {
+      updateBuffer();
+    }
+    mPos ++;
+    mBufferPos ++;
+    return BufferUtils.byteToInt(mBuffer.get());
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    checkIfClosed();
+    Preconditions.checkArgument(b != null, "Read buffer cannot be null");
+    Preconditions.checkArgument(off >= 0 && len >= 0 && len + off <= b.length,
+        String.format("Buffer length (%d), offset(%d), len(%d)", b.length, off, len));
+    if (len == 0) {
+      return 0;
+    } else if (remaining() == 0) { // End of block
+      return -1;
+    }
+
+    int toRead = (int) Math.min(len, remaining());
+    if (remainingInBuffer() > toRead) { // data is fully contained in the buffer
+      mBuffer.get(b, off, toRead);
+      mPos += toRead;
+      mBufferPos += toRead;
+      return toRead;
```

Comment:
Oh, I see. because the bytes are read from buffer rather from real input. 
So the deal is only increment counters when we really read from the input, right?
better document it somewhere